### ISSUE TITLE
when lambda  insert "__device__" "__host__",report error

### ISSUE THREAD

**Issue by lishi8168797** [microsoft/vscode-cpptools#10455]
### Environment

- OS and Version: ubuntu 20.04
- VS Code Version: 1.74.3
- C/C++ Extension Version: 1.14.0
- If using SSH remote, specify OS of remote machine: no

### Bug Summary and Steps to Reproduce

![2023-02-02 16-03-54 的屏幕截图](https://user-images.githubusercontent.com/100831761/216268106-600a0a24-7c35-4027-b496-19bbf9fbb699.png)
please see red line
it can compile,but in vscode it report error and has a red line

### Configuration and Logs

```shell
{
    "configurations": [
        {
            "name": "Linux",
            "includePath": [
                "${workspaceFolder}/**",
                "/opt/ros/noetic/include",
                "/usr/local/cuda/include"
            ],
            "defines": [],
            "compilerPath": "/usr/bin/gcc",
            "cStandard": "gnu17",
            "cppStandard": "gnu++14",
            "intelliSenseMode": "linux-gcc-x64"
        }
    ],
    "version": 4
}
```

**Comment by browntarik** [comment]
Thank you for submitting this issue! We are currently looking into it.

**Comment by browntarik** [comment]
I am having a hard time reproducing your error, could you provide a simple test case to reproduce the error? Could you also provide the error output?

**Comment by Colengms** [comment]
Hi @lishi8168797 .  It looks like you're using CUDA.  Is the code you posted in an `.cu` (or `.cuh`) file?

If you're currently using a `.cpp` file, you might try renaming it to use `.cu`.  Or, you could specify `nvcc` as the `compilerPath` to cause all `.cpp` files to be treated as CUDA.  Or, you could use a `compile_commands.json`, or a custom configuration provider, to cause files to be configured exactly as built.  The actual command line used to build the file will be considered in those cases.

Unfortunately, simply changing the language ID of the document to "CUDA C++" in the editor is insufficient to cause the C/C++ Extension to treat the file as CUDA.  I've opened https://github.com/microsoft/vscode-cpptools/issues/10494 to address that specific scenario.

**Comment by lishi8168797** [comment]
@browntarik 
![2023-02-09 10-59-47 的屏幕截图](https://user-images.githubusercontent.com/100831761/217706756-3055dd23-cdc3-47e8-bbc6-53c7e12a5b55.png)
it reported this error,and make other code after it error,in fact they are all correct.

**Comment by lishi8168797** [comment]
![2023-02-09 10-59-47 的屏幕截图](https://user-imag@Colengms s.githubusercontent.com/100831761/217707568-024295f0-fe65-4c15-9242-1b89409ea447.png)
@Colengms yes,I am using CUDA,and it is a .cu file,but it still report this red line,and make other code after it also have red line,it bother me.

**Comment by sean-mcmanus** [comment]
Can you provide the output of running the C/C++: Log Diagnostics command and/or look in that logging for an misconfiguration? I'm guessing you may need to set your compilerPath to the nvcc compiler or make sure gcc isn't being used instead.

**Comment by lishi8168797** [comment]
sorry,I may not understand that "Can you provide the output of running the C/C++",because,this code run correct,it only report this error in vscode. Does C/C++ Extension provied "insert __device__ in a lambda"?
And I set compiler with nvcc in c_cpp_properties.json,but it does not work.

**Comment by sean-mcmanus** [comment]
The C/C++: Log Diagnostics command is run from the VS Code command palette -- it can show what IntelliSense configuration is being used. It looks like something is causing the CUDA configuration to fail. @Colengms might have more info when he gets back.

**Comment by lishi8168797** [comment]
@sean-mcmanus 
Thank you,this my log:


-------- Diagnostics - 2023/2/10 10:26:20
Version: 1.14.1
Current Configuration:
{
    "name": "Linux",
    "includePath": [
        "/media/lishi/work/图像点云预处理/点云预处理/ros_env/src/pcl_preprocessing/**",
        "/opt/ros/noetic/include",
        "/usr/local/cuda/include"
    ],
    "defines": [],
    "compilerPath": "/usr/local/cuda/bin/nvcc",
    "cStandard": "gnu17",
    "cppStandard": "gnu++14",
    "intelliSenseMode": "linux-gcc-x64",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": true,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "browse": {
        "path": [
            "${workspaceFolder}/**",
            "/opt/ros/noetic/include",
            "/usr/local/cuda/include"
        ],
        "limitSymbolsToIncludedHeaders": true
    }
}
Translation Unit Mappings:
[ /media/lishi/work/图像点云预处理/点云预处理/ros_env/src/pcl_preprocessing/src/octree/octree.cu ]:
    /media/lishi/work/图像点云预处理/点云预处理/ros_env/src/pcl_preprocessing/src/octree/octree.cu
Translation Unit Configurations:
[ /media/lishi/work/图像点云预处理/点云预处理/ros_env/src/pcl_preprocessing/src/octree/octree.cu ]:
    Process ID: 7754
    Memory Usage: 553 MB
    Compiler Path: /usr/bin/gcc
    Includes:
        /opt/ros/noetic/include
        /usr/local/cuda-11.6/targets/x86_64-linux/include
        /usr/include/c++/9
        /usr/include/x86_64-linux-gnu/c++/9
        /usr/include/c++/9/backward
        /usr/lib/gcc/x86_64-linux-gnu/9/include
        /usr/local/include
        /usr/include/x86_64-linux-gnu
        /usr/include
        /media/lishi/work/图像点云预处理/点云预处理/ros_env/src/pcl_preprocessing/include
    Defines:
        __CUDA_ARCH__=520
        __CUDA_ARCH_LIST__=520
        CUDA_DOUBLE_MATH_FUNCTIONS
        __CUDACC__
        __NVCC__
        __CUDACC_VER_MAJOR__=11
        __CUDACC_VER_MINOR__=6
        __CUDACC_VER_BUILD__=55
        __CUDA_API_VER_MAJOR__=11
        __CUDA_API_VER_MINOR__=6
        __NVCC_DIAG_PRAGMA_SUPPORT__=1
    Forced Includes:
        /usr/local/cuda/include/cuda_runtime.h
    Standard Version: c++14
    IntelliSense Mode: linux-gcc-x64
    Other Flags:
        --g++
        --gnu_version=90400
        --cuda
Total Memory Usage: 553 MB

------- Workspace parsing diagnostics -------
Number of files discovered (not excluded): 38229

**Comment by sean-mcmanus** [comment]
Your logging does indicate that it's compiling in CUDA mode, so that looks like `__device__ __host__` should be recognized -- I'll wait for @Colengms to get back to see if he has any ideas what could be going wrong. (I'm guessing the /usr/bin/gcc compiler path is just showing what nvcc uses to compile the C++ code with).

**Comment by Colengms** [comment]
Hi @lishi8168797 . If your file is being correctly interpreted as CUDA, and your code compiles properly, perhaps there is an issue with IntelliSense resulting in `__host__` and `__device__` (aka `__attribute__((host))` and `__attribute__((device))`) not being recognized in certain contexts.

Can you provide a simple isolated example that we might pass along to our IntelliSense team, to further investigate the issue?

**Comment by lishi8168797** [comment]
Hi @Colengms .This is my example,thanks for your help.
=================================================
        float dim_x,dim_y,dim_z;
        float x_min,y_min_,z_min; 
        int depth_mult=1024;
        auto morton_func=[dim_x,dim_y,dim_z,depth_mult,x_min_,y_min_,z_min_]  
            __host__ __device__ 
            (const float& x,const float& y,const float& z,int& code){
            int cellx,celly,cellz;
            cellx = min((int)std::floor(depth_mult * (x-x_min_)/dim_x), depth_mult - 1);
            celly = min((int)std::floor(depth_mult * (y-y_min_)/dim_y), depth_mult - 1);
            cellz = min((int)std::floor(depth_mult * (z-z_min_)/dim_z), depth_mult - 1);
            //......................9876543210
            cellx = (cellx | (cellx << 10)) & 0x000f801f; //............98765..........43210
            cellx = (cellx | (cellx <<  4)) & 0x00e181c3; //........987....56......432....10
            cellx = (cellx | (cellx <<  2)) & 0x03248649; //......98..7..5..6....43..2..1..0
            cellx = (cellx | (cellx <<  2)) & 0x09249249; //....9..8..7..5..6..4..3..2..1..0
            //......................9876543210
            celly = (celly | (celly << 10)) & 0x000f801f; //............98765..........43210
            celly = (celly | (celly <<  4)) & 0x00e181c3; //........987....56......432....10
            celly = (celly | (celly <<  2)) & 0x03248649; //......98..7..5..6....43..2..1..0
            celly = (celly | (celly <<  2)) & 0x09249249; //....9..8..7..5..6..4..3..2..1..0
            celly=celly << 1;
            //......................9876543210
            cellz = (cellz | (cellz << 10)) & 0x000f801f; //............98765..........43210
            cellz = (cellz | (cellz <<  4)) & 0x00e181c3; //........987....56......432....10
            cellz = (cellz | (cellz <<  2)) & 0x03248649; //......98..7..5..6....43..2..1..0
            cellz = (cellz | (cellz <<  2)) & 0x09249249; //....9..8..7..5..6..4..3..2..1..0
            cellz=cellz << 2;
            code=cellx | celly | cellz;
        };

        thrust::for_each(thrust::device,thrust::make_zip_iterator(x_ptr,y_ptr,z_ptr,temp_code.begin()),
        thrust::make_zip_iterator(x_ptr+size,y_ptr+size,z_ptr+size,temp_code.begin()+size),
        thrust::make_zip_function(morton_func));

**Comment by Colengms** [comment]
Hi @lishi8168797 .  I've opened an internal issue (1746549) against VS, as this issue is with the IntelliSense parser which is shared as VS.

**Comment by sean-mcmanus** [comment]
The fix is available with 1.15.0 (pre-release): https://github.com/microsoft/vscode-cpptools/releases/tag/v1.15.0

**Comment by lishi8168797** [comment]
> The fix is available with 1.15.0 (pre-release): https://github.com/microsoft/vscode-cpptools/releases/tag/v1.15.0

It works well ! Thank you very much !
